rust/datafusion/src/physical_plan/mod.rs

```diff
@@ -100,6 +100,30 @@ pub enum Distribution {
     SinglePartition,
 }
 
+/// Represents the result from an expression
+pub enum ColumnarValue {
```

Comment:
Perhaps we could push the new `ColumnarValue` enum down to the core arrow crate since it isn't specific to DataFusion?

Reply:
possibly; where / how could you see the `ColumnarValue` enum used in core arrow? also wouldn't `ScalarValue` need to move as well?

Reply:
C++ uses a `Datum`, which is also an enum over Scalar, Array, and a few other things.
We could have a separate module called `arrow::scalar`, then in the long run we could convert the compute kernels to take `Datum`, and push the optimisations of "array vs scalar" there

Reply:
I agree.

It would also simplify the API of many vertical operations (e.g. aggregates), as their input and result would have a common type. I have a branch on which I am doing that, but I have not finish it yet.

Reply:
I also went down that rabbit hole over a year ago, yeah it's not easy

Reply:
I agree with @alamb ; I think if we wanted to move `ColumnarValue`  into `arrow` it would be better to do that in a separate PR after this one has been merged

Reply:
Note that our use of `Datum` in C++ is **far from perfect**, it's actually a relatively complex data structure and I've contemplated using something simplified that doesn't have as many non-trivial C++ objects inside it in the internals of function execution